tensorboard/manager_e2e_test.py

```diff
@@ -0,0 +1,302 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""End-to-end tests for `tensorboard.manager`."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import datetime
+import errno
+import json
+import os
+import pipes
+import signal
+import subprocess
+import tempfile
+import textwrap
+
+import six
+from six.moves import urllib
+import tensorflow as tf
+
+try:
+  # python version >= 3.3
+  from unittest import mock  # pylint: disable=g-import-not-at-top
+except ImportError:
+  import mock  # pylint: disable=g-import-not-at-top,unused-import
+
+from tensorboard import manager
+
+
+class ManagerEndToEndTest(tf.test.TestCase):
+
+  def setUp(self):
+    super(ManagerEndToEndTest, self).setUp()
+
+    # Spy on subprocesses spawned so that we can kill them.
+    self.popens = []
+    class PopenSpy(subprocess.Popen):
+      def __init__(p, *args, **kwargs):
+        super(PopenSpy, p).__init__(*args, **kwargs)
+        self.popens.append(p)
+    popen_patcher = mock.patch.object(subprocess, "Popen", PopenSpy)
+    popen_patcher.start()
+
+    # Make sure that temporary files (including .tensorboard-info) are
+    # created under our purview.
+    self.tmproot = os.path.join(self.get_temp_dir(), "tmproot")
+    os.mkdir(self.tmproot)
+    tmpdir_environ = {"TMPDIR": self.tmproot}
+    tmpdir_environ_patcher = mock.patch.dict(os.environ, tmpdir_environ)
+    tmpdir_environ_patcher.start()
+    self.addCleanup(tmpdir_environ_patcher.stop)
+    tempfile.tempdir = None  # force `gettempdir` to reinitialize from env
+    self.assertEqual(tempfile.gettempdir(), self.tmproot)
+    self.info_dir = manager._get_info_dir()  # ensure that directory exists
+
+    # Add our Bazel-provided `tensorboard` to the system path.
+    tensorboard_binary_dir = os.path.realpath("./tensorboard/")
+    path_environ = {
+        "PATH": os.pathsep.join((tensorboard_binary_dir, os.environ["PATH"])),
+    }
+    path_environ_patcher = mock.patch.dict(os.environ, path_environ)
+    path_environ_patcher.start()
+    self.addCleanup(path_environ_patcher.stop)
+    self._ensure_tensorboard_on_path(tensorboard_binary_dir)
+
+  def tearDown(self):
+    failed_kills = []
+    for p in self.popens:
+      try:
+        p.kill()
+      except Exception as e:
+        if isinstance(e, OSError) and e.errno == errno.ESRCH:
+          # ESRCH 3 No such process: e.g., it already exited.
+          pass
+        else:
+          # We really want to make sure to try to kill all these
+          # processes. Continue killing; fail the test later.
+          failed_kills.append(e)
+    for p in self.popens:
+      p.wait()
```

Comment:
Is there a way to forcibly kill the process? Does this mean, in the worst case scenario, we will wait ~15s (except for the timeout test)?

Reply:
The preceding `for`-loop will kill the processes. On Unix it sends
`SIGKILL`; on Windows it sends `TerminateProcess`. You can’t ignore,
block, or recover from those signals. I don’t see any way that these
kills could fail (and I never saw the `failed_kills` check actually
fail).

Reaping them with `wait` is just good hygiene; it frees up the child’s
slot in the process table.

Reply:
Ok, thanks for the explanation!